IMDA-BTG/aiverify
# Feature/issue 53/canvas dynamic height

## Description

Implements feature request https://github.com/IMDA-BTG/aiverify/issues/53. This feature allows developers to create widgets with dynamic height.

Changes:
- Add new property dynamicHeight to widget meta schema.
- Update [puppeteer](https://www.npmjs.com/package/puppeteer) dependency in ai-verify-apigw project.
- In canvas, dynamic height widgets will not be able to drag or drop above other widgets. Widget height will auto expand to bottom of canvas.
- In report module, manually layout the widgets on report page instead of using react-grid-layout layout; to cater for dynamic height widgets.

After update of code, need to restart portal and apigw projects.

For ai-verify-apigw, need to run the following before restarting:
```
cd ai-verify-apigw
npm i
```

## Motivation and Context

See https://github.com/IMDA-BTG/aiverify/issues/53.

## Type of Change













feat: Dynamic Height for Canvas Widgets

## How to Test

Step 1) Create a new plugin with a widget. The widget should include meta property with dynamicHeight set to true.
For example, a test widget with cid `testdh`.
`testdh.meta.json`
```
{
  "cid": "testdh",
  "widgetSize": {
    "minW": 12,
    "minH": 3,
    "maxW": 12,
    "maxH": 36
  },
  "name": "Test Dynamic Height",
  "dynamicHeight": true
}
```
`testdh.mdx`
```
export const numBlocks = 20;

<table style={{ width:'100%', borderCollapse:'collapse' }}>
{[...Array(numBlocks)].map((x, i) => (
  <tr style={{ breakInside: 'avoid', border:'1px solid black' }}>
    <td key={`block-${i}`} style={{ height:'200px', verticalAlign:'middle', textAlign:'center' }}>{i}</td>
  </tr>
))}
</table>
```

Step 2) Install the plugin in the ai-verify-portal.

Step 3) Create new blank project and add the test widget testdh to the canvas.

Step 4) Generate report.

Step 5) View report and check that the height of the widgets has expanded over multiple pages, with the right page breaks.


## Checklist

Please check all the boxes that apply to this pull request using "x":


## Screenshots (if applicable)

[If the changes involve visual modifications, include screenshots or GIFs that demonstrate the changes.]

## Additional Notes

[Add any additional information or context that might be relevant to reviewers.]
